MC/Core/Src/freertos.c

```diff
@@ -0,0 +1,390 @@
+
+	}
+}
+
+void MC_ReceiveEndThread(void *argument) {
+
+	CAN_Frame_t end_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER_K2); //Initialize CAN_Frame to send to MC to receive second data frame
+
+	end_frame.id_type = CAN_ID_EXT;
+	end_frame.data_length = 0;
+	end_frame.rtr = 1;
+
+	uint8_t received = 0;
+
+	for(;;){
+
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan3)) {
+			CAN_send_frame(end_frame);
+		}
+
+		if(HAL_CAN_GetRxFifoFillLevel(&hcan3, CAN_RX_FIFO0)){
+			end_frame = CAN_get_frame(&hcan3, CAN_RX_FIFO0);
+			received = 1;
+		}
+
+
+		if(received){
+			osMessageQueuePut(end_frame_q, &end_frame, 0, osWaitForever);
+			osSemaphoreRelease(MCSemEndHandle);
+
+		}
+
+	}
+}
+
+
+
+void MC_SendStartThread(void *argument){
+
+	CAN_Frame_t start_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER_K1); //Initialize CAN_frame to send to MC
+	CAN_Frame_t send_frame = CAN_frame_init(&hcan1, MOTOR_CONTROLLER_K1); //Initialize CAN_frame to send to RPI
+
+	Motor_Controller_t mc_data = {0};
+
+	for(;;){
+
+		osSemaphoreAcquire(MCSemStartHandle, osWaitForever);
+
+		osMessageQueueGet(start_frame_q, &start_frame, NULL, osWaitForever);
+
+		//Process data from MC
+		mc_data.direction = direction(CAN_get_segment(start_frame, DRIVING_DIRECTION_K));
+		mc_data.speed = CAN_get_segment(start_frame, MOTOR_SPEED_K);
+		mc_data.error_code = error_code(CAN_get_segment(start_frame, MOTOR_ERROR_CODE_K));
+
+		//Create new CAN_frame with proccessed data
+		CAN_set_segment(&send_frame, DRIVING_DIRECTION_K, mc_data.direction);
+		CAN_set_segment(&send_frame, MOTOR_SPEED_K, mc_data.speed);
+		CAN_set_segment(&send_frame, MOTOR_ERROR_CODE_K, mc_data.error_code);
+
+	    //Send data to RPI
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1)) {
+	        CAN_send_frame(send_frame);
+	    }
+
+
+	    osDelay(1);
+
+
+	}
+}
+
+void MC_SendEndThread(void *argument){
+
+	CAN_Frame_t end_frame = CAN_frame_init(&hcan3, MOTOR_CONTROLLER_K2);
+	CAN_Frame_t send_frame = CAN_frame_init(&hcan1, MOTOR_CONTROLLER_K2);
+
+	Motor_Controller_t mc_data = {0};
+
+	for(;;){
+
+		osSemaphoreAcquire(MCSemEndHandle, osWaitForever);
+
+		osMessageQueueGet(end_frame_q, &end_frame, NULL, osWaitForever);
+
+
+		mc_data.voltage = process_bits(CAN_get_segment(end_frame, BATTERY_VOLTAGE_K));
+		mc_data.current = process_bits(CAN_get_segment(end_frame, BATTERY_CURRENT_K));
+		mc_data.motor_temp = process_bits(CAN_get_segment(end_frame, MOTOR_TEMP_K));
+		mc_data.controller_temp = process_bits(CAN_get_segment(end_frame, MOTOR_CONTROLLER_TEMP_K));
+
+
+		CAN_set_segment(&send_frame, BATTERY_VOLTAGE_K, mc_data.voltage);
+		CAN_set_segment(&send_frame, BATTERY_CURRENT_K, mc_data.current);
+		CAN_set_segment(&send_frame, MOTOR_TEMP_K, mc_data.motor_temp);
+		CAN_set_segment(&send_frame, MOTOR_CONTROLLER_TEMP_K, mc_data.controller_temp);
+
+	    if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan1)) {
+	    	CAN_send_frame(send_frame);
+	    }
+
+	    osDelay(1);
+
+
+	}
+}
+
+DrivingDirection direction(uint8_t data){
```

Comment:
Not sure if you wrote these functions or if they're from old code but we might want function names to be a bit more descriptive for readability

Reply:
also if you can comment out your code, that would also be a huge help to readability. I understand bc I worked on the sensor module rtos but it would be hard for those unfamiliar

Reply:
Okay wait some parts are commented and others aren't, but you get what I mean

Reply:
Changed function name to bits_to_direction() and added comments
Done